proxy/src/main/java/ru/leymooo/botfilter/captcha/CaptchaGeneration.java

```diff
@@ -12,109 +14,97 @@
-        }
-        CachedCaptcha.generated = true;
-        executor.shutdownNow();
-        System.gc();
-        BungeeCord.getInstance().getLogger().log( Level.INFO, "[BotFilter] Капча сгенерированна за {0} мс", System.currentTimeMillis() - start );
-    }
 
+            List<CachedCaptcha.CaptchaHolder> holders = Collections.synchronizedList( new ArrayList<>() );
 
-    private Color randomNotWhiteColor()
-    {
-        Color color = MapPalette.colors[rnd.nextInt( MapPalette.colors.length )];
+            for ( int i = 1; i <= captchaCount; i++ )
+            {
+                executor.execute( new CaptchaGenerationTask( executor, fonts, holders ) );
+            }
 
-        int r = color.getRed();
-        int g = color.getGreen();
-        int b = color.getBlue();
+            long start = System.currentTimeMillis();
+            ThreadPoolExecutor ex = (ThreadPoolExecutor) executor;
+            while ( ex.getActiveCount() != 0 )
+            {
+                //Отображаем прогресс генерации только после полной загрузки банжи, чтобы логи не путались с другими важными логами при включении
+                if ( BungeeCord.getInstance().isEnabled() )
+                {
+                    BungeeCord.getInstance().getLogger().log( Level.INFO, "[BotFilter] Генерирую капчу [" + ( captchaCount - ex.getQueue().size() ) + "/" + captchaCount + "]" );
+                }
 
-        if ( r == 255 && g == 255 && b == 255 )
-        {
-            return randomNotWhiteColor();
-        }
-        if ( r == 220 && g == 220 && b == 220 )
-        {
-            return randomNotWhiteColor();
-        }
-        if ( r == 199 && g == 199 && b == 199 )
-        {
-            return randomNotWhiteColor();
-        }
-        if ( r == 255 && g == 252 && b == 245 )
-        {
-            return randomNotWhiteColor();
-        }
-        if ( r == 220 && g == 217 && b == 211 )
-        {
-            return randomNotWhiteColor();
-        }
-        if ( r == 247 && g == 233 && b == 163 )
-        {
-            return randomNotWhiteColor();
-        }
-        return color;
-    }
+                //Текущий список всех капч каждую секунду вставляем в новый список для гарантии потокобезопасности.
+                PacketUtils.captchas.setCaptchas( new ArrayList<>( holders ) );
+                try
+                {
+                    Thread.sleep( 1000L );
+                } catch ( InterruptedException ex1 )
+                {
+                    BungeeCord.getInstance().getLogger().log( Level.WARNING, "[BotFilter] Не могу сгенерировать капчу. Выключаю банджу", ex1 );
+                    System.exit( 0 );
+                    return;
+                }
+            }
 
-    private String randomAnswer()
-    {
-        if ( rnd.nextBoolean() )
+            executor.shutdownNow();
+
+            //Окончательно устанавливаем оставшиеся капчи
+            PacketUtils.captchas.setCaptchas( new ArrayList<>( holders ) );
+            System.gc();
```

Comment:
I wouldn't all the GC tbh

Reply:
> I wouldn't all the GC tbh

The captcha generation behavior is identical to how it was, including the call to System.gc(), and there is nothing bad about it, in fact, a lot of memory is consumed during generation, it's not a problem to always call the garbage collector at the end

Reply:
I don't like the idea in general: https://stackoverflow.com/questions/5086800/java-garbage-collection